reference_submissions/librispeech_conformer/librispeech_jax/submission.py

```diff
@@ -0,0 +1,186 @@
+"""Training algorithm track submission functions for LibriSpeech."""
+from typing import Dict, Iterator, List, Tuple
+
+import jax.numpy as jnp
+import jax
+import flax 
+import optax
+from flax import jax_utils
+import functools
+import jax.lax as lax
+import numpy as np
+from absl import logging
+import operator
+from algorithmic_efficiency import spec
+
+_GRAD_CLIP_EPS = 1e-6
+
+def get_batch_size(workload_name):
+  # Return the global batch size.
+  del workload_name
+  return 128
+
+
+def create_learning_rate_fn(hparams: spec.Hyperparameters):
+  """Create learning rate schedule."""
+  base_learning_rate = hparams.base_lr
+  warmup_fn = optax.linear_schedule(
+      init_value=0.,
+      end_value=base_learning_rate,
+      transition_steps=hparams.warmup_steps)
+  cosine_steps = max(hparams.num_training_steps - hparams.warmup_steps, 1)
+  cosine_fn = optax.cosine_decay_schedule(
+      init_value=base_learning_rate,
+      decay_steps=cosine_steps)
+  schedule_fn = optax.join_schedules(
+      schedules=[warmup_fn, cosine_fn],
+      boundaries=[hparams.warmup_steps])
+  return schedule_fn
+
+def optimizer(hyperparameters: spec.Hyperparameters, num_train_examples: int):
+  learning_rate_fn = create_learning_rate_fn(hyperparameters)
+  opt_init_fn, opt_update_fn = optax.adamw(
+      b1=hyperparameters.beta1,
+      b2=hyperparameters.beta2,
+      eps=hyperparameters.epsilon,
+      learning_rate=learning_rate_fn)
+  return opt_init_fn, opt_update_fn
+
+
+def init_optimizer_state(workload: spec.Workload,
+                         model_params: spec.ParameterContainer,
+                         model_state: spec.ModelAuxiliaryState,
+                         hyperparameters: spec.Hyperparameters,
+                         rng: spec.RandomState, log_dir = None) -> spec.OptimizerState:
+  model_params
+  del model_state
+  del rng
+  workload.create_summary_writer(log_dir)
```

Comment:
I don't think we should do this here, I think we should move it to submission_runner.py and before we call `with profiler.profile('Initializing optimizer'):` we should have:
```
if log_dir:
  workload.create_summary_writer(log_dir)
```

(and delete `log_dir` from the args)

Reply:
great suggestion!